GC: Detect references added via attach ops

markfields opened the pull request:
## Description

Fixes [AB#5757](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/5757).  Follow-on to #18456 

With this PR we fill a gap in detecting references as they come - in the case where an object is re-referenced in a new object (the first reference appears in that attach op). _Background: This is important in the case where that reference is quickly removed (before the next GC run). Otherwise we would not update the unreferenced timestamp and GC calculations would be incorrect._

### Implementation

Just as we include the container's GC state in each summary, we need to include an object's initial GC Data (list of references) in its attach summary.  This data can be parsed out of the attach message on process and GC can be notified.

### Testing

Existing coverage:

- GC reference detection for DataStore attach: `gcUnreferencedTimestamp.spec.ts`
- Overall attach flow (op generation and processing), including compat between runtimes: `attachLifecycle.spec.ts`

Additional coverage:

- `gcUnreferencedTimestamp.spec.ts`:
  - [x] Use a DDS in the chain
  - [x] Use multiple DDSes in the chain
- `gcInactiveNodes.spec.ts` - Revival happens on local client that sent the attach op
  - [x] DataStore attach
  - [x] DDS attach

## Reviewer Guidance

Question: Better to split this into more than one PR?  e.g. there is some light refactoring to how attach messages are prepared that could come first, or DDS attach and DataStore attach could come in 2 steps.

Test coverage considered but not planning to do:

- Test outbound routes to attachment blobs.  Similar enough to outbound routes to DataStores
- Test that revival on local client happens for Summarizer client. No interesting divergence from Interactive client codepath, and it violates the "no local changes in Summarizer" tenet that is about to get harder and harder to exercise as we enable the new code by default

markfields committed:
Compute GC data during DataStore/DDS realize based on initial attach state only

- Do it only when loading from an attach snapshot
- Do it before processing pending ops for the node

markfields committed:
Unblock CI

markfields committed:
Revert temp test change and cherry-pick launch.json fix

markfields committed:
Only get GC Data if GC is enabled (ish)

markfields committed:
Fix hack (learn! jk)

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -452,6 +456,32 @@ export abstract class FluidDataStoreContext
 		assert(channel !== undefined, 0x140 /* "undefined channel on datastore context" */);
+
+		//* TODO: Properly plumb this bit from ContainerRuntime
+		const gcEnabled = (this._containerRuntime as any).garbageCollector?.shouldRunGC ?? true; //* Some tests mock CR and this isn't there.
```
I wonder what percentage of attach ops are realized immediately after they arrive. I assume in the summarizer client, that doesn't happen very frequently.

I really don't like that this has to be in the realize function. I would prefer a synchronous process over an asynchronous "processing" of the attach op.

tyler-cai-microsoft replied:
Another side note, does the local datastore context need this functionality? Technically it's already attached and we could just run getGCData on it. Just a thought.

markfields replied:
> does the local datastore context need this functionality

Had to think about it for a minute.

Summarizer is the main customer here, and it "shouldn't" have local changes.  But it can today, so in that case we need to do something.  Need to look into how the attach flow works for local things when the attach op is ack'd

(Interactive clients benefit as well, in that they clear unreferenced state when notified of a new reference, which reduces false positives of warning logs.)

markfields replied:
@agarwal-navin - thoughts on local channels? I need to address this in either case. I'll look into it soon but if you already can see what needs to be done, feel free to chime in :)

markfields replied:
Resolving this thread.

- Switched to the snapshot-based approach rather than getting GC data during realize
- With that, we are pulling the GC Data from the snapshot synchronously when the attach op is processed, so we can easily do that before branching off for local v. remote messages.

markfields committed:
Add GC Data to DDS and DataStore attach summary.

It's in a .gcdata blob
Hopefully I got the GC Data Builder right

markfields committed:
DataStoreRuntime: Read GC Data from snapshot and compare with what's computed.

markfields commented on packages/runtime/datastore/src/dataStoreRuntime.ts:
```diff
@@ -864,7 +875,14 @@ export class FluidDataStoreRuntime
+		// We need to include the GC Data so remote clients can learn of this DataStore's outbound routes
+		this.updateGCNodes(gcDataBuilder);
+		const gcData = gcDataBuilder.getGCData();
+		addBlobToSummary(attachSummary, gcDataBlobKey, JSON.stringify(gcData));
```
@agarwal-navin  interested in your feedback on this bit of code (everything in this function)

markfields commented on packages/runtime/datastore/src/remoteChannelContext.ts:
```diff
@@ -94,6 +94,26 @@ export class RemoteChannelContext implements IChannelContext {
 				this.id,
 			);
 
+			//* TODO: Read from snapshot instead (like in DataStoreRuntime)
```
Highlighting this - Strategy 2 is not fully implemented. (And I don't think we have test coverage of DDS attach scenario anyway yet)

agarwal-navin commented:
I am having a hard time reviewing this because I can't tell which part is needed for which strategy.

markfields committed:
Some notes from talking with Navin

markfields committed:
Fix temp "gut check" assert to confirm no hits in CI before proceeding.

markfields committed:
Switch log event gcUnknownOutboundReferences from Error to Generic

markfields committed:
Generate DS attach message in ContaineRuntime not DataStoreContext.

Achieves symmetry between DS attach and DDS attach.
Introduces a new function on DSRuntime for getting all the datat needed for attach

markfields committed:
Read .gcdata blob synchronously when processing the attach message

markfields committed:
Read .gcdata blob synchronously when processing the attach message (DDS)

markfields committed:
Merge branch 'main' into gc/ref-via-attach-3

markfields committed:
Address some todos

markfields committed:
CI fix

markfields committed:
Address some more todo's

markfields committed:
Get DDS test to actually cover DDS codepath

Before, due to the chaining, C's attach op was triggering D timestamp reset apart from anything with DDS B.1

markfields committed:
Another DDS test

markfields committed:
Fix tests

markfields committed:
Disable DataStore id test hack to let CI pass

markfields committed:
Add local attach op test (Interactive client)

markfields committed:
Cover DDS and DataStore attach op in InactiveObject test

markfields committed:
Remove some //*

markfields committed:
Deal with most of //* comments.  Now tests are failing

markfields committed:
Remove compat for gcUnreferencedTimestamp tests

It was useful for the last PR to make sure initial implementation worked.
Now it's trouble because in compat scenarios there will be a gap, and it's ok.
Setting min version.

markfields committed:
Remove log since it'll get logged too much

I could update the sampled telemetry helper to support that but going to just remove this log since the scenario is very niche

markfields committed:
Merge branch 'main' into gc/ref-via-attach-3

markfields committed:
build fix

markfields committed:
nits

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -665,6 +665,13 @@ export abstract class FluidDataStoreContext
 		}
 	}
 
+	public addedGCOutboundRoute(fromPath: string, toPath: string) {
```
Nit: `addedGCOutboundReference` so everything matches

markfields replied:
I named it differently on purpose - it has a different signature.  I'll make sure there's comments everywhere to highlight that.

tyler-cai-microsoft commented on packages/runtime/datastore/src/dataStoreRuntime.ts:
```diff
@@ -482,7 +486,7 @@ export class FluidDataStoreRuntime
 		this.notBoundedChannelContextSet.delete(channel.id);
 		// If our data store is attached, then attach the channel.
 		if (this.isAttached) {
-			this.attachChannel(channel);
```
Our attach flow is so complicated, I don't know what implications this change does.

markfields replied:
Good news - it's just a rename :) For more symmetry with other layers

tyler-cai-microsoft commented on packages/runtime/runtime-definitions/src/dataStoreContext.ts:
```diff
@@ -486,12 +496,23 @@ export interface IFluidDataStoreContext
 	 * @param srcHandle - The handle of the node that added the reference.
 	 * @param outboundHandle - The handle of the outbound node that is referenced.
 	 */
 	addedGCOutboundReference?(srcHandle: IFluidHandle, outboundHandle: IFluidHandle): void;
```
I wonder if this can be overloaded

markfields replied:
Yeah, I thought about that a bit. Expanding the "in" types means someone _using_ this interface sees that as a backwards compatible change (existing calls are fine) but someone _implementing_ the interface sees it as a non-back-compat change. Of course no one implements this, and our type tests don't even cover function signature changes, but I'd rather keep it simple.

tyler-cai-microsoft commented on packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts:
```diff
@@ -21,19 +21,14 @@ import {
- * (skip all other compat types)
  */
-describeCompat("GC unreferenced timestamp", "FullCompat", (getTestObjectProvider, apis) => {
+describeCompat("GC unreferenced timestamp", "2.0.0-rc.1.0.0", (getTestObjectProvider, apis) => {
```
For compatibility sake, I'm assuming the attach ops should still be able to be processed by older versions. Especially now that we may get some much bigger compatibility swings.

markfields replied:
Yes the new attach ops should be. I didn't think about test coverage for this, it seems vacuously true (just adding a blob). I'll consider it...

markfields replied:
I did confirm once that running the bulk of these tests under FullCompat works fine.  I'm leaving this as NoCompat for now, see other thread in this file about compat

tyler-cai-microsoft commented on packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts:
```diff
@@ -659,14 +641,148 @@ describeCompat("GC unreferenced timestamp", "FullCompat", (getTestObjectProvider
+			 *
+			 * Validates that the unreferenced time for C is t2 which is > t1.
+			 */
+			it(`Scenario 8 - Node re-referenced via new unreferenced DDS attaching`, async () => {
```
Another option would be to move these scenarios under a different `describeCompat` or boundary so we don't lose any of our original coverage.

markfields replied:
I spent a couple min thinking about this, it's a bit annoying to do since there's so much shared helpers / globals within that top-level describe.  Also it's an awkward time since specifying min version is broken. So I'm going to let it go for now, knowing that we'll get min version coverage for rc.1.0 onward when that's fixed.

tyler-cai-microsoft commented on packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts:
```diff
@@ -659,14 +641,148 @@ describeCompat("GC unreferenced timestamp", "FullCompat", (getTestObjectProvider
+			 *
+			 * Validates that the unreferenced time for C is t2 which is > t1.
+			 */
+			it(`Scenario 9 - Node re-referenced via new unreferenced DDS attaching (extra indirection)`, async () => {
```
Do we test scenarios where dataStoreB has a reference to dataStoreC via dataStoreB's ddsB and is not yet attached, and then dataStoreB is attached which means the attach message would contain both dataStoreB's and ddsB's gcData?

markfields replied:
I believe that's Scenario 6. `dataStoreB` is detached, its `_root` DDS has a reference to `dataStoreC`. B's attach message includes GC data for the whole dataStore (including `_root` DDS) like you describe.  I'll step through that test one more time to confirm.

markfields committed:
PR feedback

markfields committed:
Merge remote-tracking branch 'origin/main' into gc/ref-via-attach-3

markfields committed:
Split into separate functions getting Attach GC Data from getting Attach Summary

markfields committed:
Resolve //* notes to self

agarwal-navin commented on packages/runtime/container-runtime/src/dataStores.ts:
```diff
@@ -195,11 +202,66 @@ export class DataStores implements IDisposable {
+		// Allows longitudinal tracking of various state (e.g. foundGCData), and some sampled details
+		if (this.shouldSendAttachLog) {
+			this.shouldSendAttachLog = false;
+			this.mc.logger.sendTelemetryEvent({
```
Should you log this only if `foundGCData` is true? So that we dont end up not getting the data for that case.

markfields replied:
I don't care to see every single hit here. The hypothesis is that over time as Loop deploys the bump with this change, we'll see more and more and more with foundGCData true: I don't expect to do much more analysis beyond that, and don't want to log on every attach x N collaborators

markfields replied:
Circling back - I _do_ want some logs with `foundGCData` false.  They will start out as the majority but then get more and more rare.

agarwal-navin commented on packages/runtime/datastore/src/dataStoreRuntime.ts:
```diff
@@ -572,6 +576,31 @@ export class FluidDataStoreRuntime
 	}
 
+	/** @returns true if it found/processed GC Data, false otherwise */
+	private extractAndProcessAttachGCData(id: string, entries: ITreeEntry[]): boolean {
```
This can be a helper function outside of the class and can be re-used in both the call sites right?

agarwal-navin replied:
That way they won't get out of sync.

markfields replied:
Yeah I was waiting for someone to say that. I can do it. The signatures of the GC function differ but I can deal with it

agarwal-navin commented on packages/runtime/datastore/src/dataStoreRuntime.ts:
```diff
@@ -812,6 +849,61 @@ export class FluidDataStoreRuntime
 
 	public getAttachSummary(telemetryContext?: ITelemetryContext): ISummaryTreeWithStats {
+		const summaryBuilder = new SummaryTreeBuilder();
+		this.getAttachData((contextId: string, context: LocalChannelContextBase) => {
```
I got really confused how this uses `this.getAttachData` which doesn't really return anything but puts sstuff in the summaryBuilder.
It would be lot cleaner and readable to just duplicate this code in the 2 functions.

markfields replied:
Good feedback.  I renamed it to make the interaction clearer.  It's kind of a visitor pattern.

agarwal-navin commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -1083,7 +1121,17 @@ export class LocalFluidDataStoreContextBase extends FluidDataStoreContext {
+	/**
+	 * @see FluidDataStoreContext.getAttachData
+	 */
+	public getAttachData(
```
It would be cleaner to have 2 separate functions - getAttachSummary and getAttachGCData. It keeps them summary and GC separate and you won't need the `includeGCData` param here.

markfields replied:
I was debating - It's minimal but I like avoiding doubling up on those asserts.  Felt worth it - and the function is factored so it will be trivial to pull apart later if needed.

markfields replied:
(Decided to Won't Fix this)

agarwal-navin commented on packages/runtime/datastore/src/dataStoreRuntime.ts:
```diff
@@ -572,6 +576,31 @@ export class FluidDataStoreRuntime
 	}
 
+	/** @returns true if it found/processed GC Data, false otherwise */
+	private extractAndProcessAttachGCData(id: string, entries: ITreeEntry[]): boolean {
```
Consider adding a comment about what this function does. It's not very apparent :)

agarwal-navin reviewed (approved):
LGTM. I left some comments and didn't look closely at tests

markfields committed:
Extract shared helper function; some nits

markfields committed:
Renames for clarity

markfields committed:
Merge remote-tracking branch 'origin/main' into gc/ref-via-attach-3

markfields committed:
ultra nit
